core/src/main/java/com/linecorp/armeria/server/cors/AbstractCorsPolicyBuilder.java

```diff
@@ -0,0 +1,383 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.cors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.google.common.base.Ascii;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.server.cors.CorsConfig.ConstantValueSupplier;
+
+import io.netty.util.AsciiString;
+
+/**
+ * Contains information for the build of the CORS policy.
+ *
+ * @see ChainedCorsPolicyBuilder
+ * @see CorsPolicyBuilder
+ */
+@SuppressWarnings("rawtypes")
+abstract class AbstractCorsPolicyBuilder<B extends AbstractCorsPolicyBuilder> {
+    private final Set<String> origins;
+    private boolean credentialsAllowed;
+    private boolean nullOriginAllowed;
+    private long maxAge;
+    private final Set<AsciiString> exposedHeaders = new HashSet<>();
+    private final Set<HttpMethod> allowedRequestMethods = new HashSet<>();
+    private final Set<AsciiString> allowedRequestHeaders = new HashSet<>();
+    private final Map<AsciiString, Supplier<?>> preflightResponseHeaders = new HashMap<>();
+    private boolean preflightResponseHeadersDisabled;
+
+    AbstractCorsPolicyBuilder() {
+        origins = Collections.emptySet();
+    }
+
+    AbstractCorsPolicyBuilder(String... origins) {
+        requireNonNull(origins, "origins");
+        for (int i = 0; i < origins.length; i++) {
+            if (origins[i] == null) {
+                throw new NullPointerException("origins[" + i + ']');
+            }
+        }
+        this.origins = Arrays.stream(origins).map(Ascii::toLowerCase).collect(toImmutableSet());
+    }
+
+    @SuppressWarnings("unchecked")
+    B self() {
+        return (B) this;
+    }
+
+    public Set<String> origins() {
+        return origins;
+    }
+
+    /**
+     * Determines if the policy allows a null origin.
+     */
+    boolean isNullOriginAllowed() {
+        return nullOriginAllowed;
+    }
+
+    /**
+     * Web browsers may set the 'Origin' request header to 'null' if a resource is loaded
+     * from the local file system. Calling this method will enable a successful CORS response
+     * with a {@code "null"} value for the the CORS response header 'Access-Control-Allow-Origin'.
```

Comment:
nit: `"null"` -> `null`, `'Access-Control-Allow-Origin'` -> `{@code Access-Control-Allow-Origin}`?

Reply:
I'm not sure about that because it literally means a String value "null". I think {@code null} may be more confusing. I also think that 'Access-Control-Allow-Origin' may be better with quotes for disambiguation.